docs/design/proposals/codegen-refactor/centralize_1.md

```diff
@@ -0,0 +1,197 @@
+
+```
+
+* Note, for scenarios that do both `reconcile` and output Go code (ex: [late_initialize::FindLateInitializedFieldNames](https://github.com/aws-controllers-k8s/code-generator/blob/25c43e827527b43c652b6e1995265c8d6027567f/pkg/generate/code/late_initialize.go#L31), these will need to be refactored into separate pieces prior to migration
+
+
+#### Align access/interface for `__ForStruct` funcs
+The goal is to make it easier for clients to use `SetResource` and `SetSDK` by removing the need to know the target type.
+
+1. Refactor access (public vs. private) for `SetResourceFor___`, and `SetSDKFor___`
+  * `setResourceForContainer` and `setSDKForContainer` are private while `SetResourceForStruct` and `SetSDKForStruct` are public despite the former pair being more general than the latter. Access should be swapped so users calling this in [controller.go](https://github.com/aws-controllers-k8s/code-generator/blob/25c43e827527b43c652b6e1995265c8d6027567f/pkg/generate/ack/controller.go#L102) don't need to worry about Struct or what type.. just pass container. Leaves it open-ended for override types too.
+  * `GoCodeCompare` vs `GoCodeCompareStruct` .. can [these](https://github.com/aws-controllers-k8s/code-generator/blob/25c43e827527b43c652b6e1995265c8d6027567f/pkg/generate/ack/controller.go#L119-L122) be consolidated?
+
+```
+// Current
+SetResourceForStruct()
+setResourceForContainer()
+setResourceForSlice()
+setResourceForMap()
+setResourceForScalar()
+
+// access in controller.go
+"GoCodeSetResourceForStruct" : ... return code.SetResourceForStruct()
+
+// Proposed
+SetResourceForContainer()
+setResourceForStruct()
+setResourceForSlice()
+setResourceForMap()
+setResourceForScalar()
+
+// access in controller.go
+"GoCodeSetResourceForContainer" : ... return code.SetResourceForContainer()
+
+```
+
+
+#### In Scope
+* Centralize code related to `reconciliation` to `model` package
+* Expose Getters in `config` and replace previous helpers/wrappers
+* Update public/private interfaces so users need only focus on the General case and not be bogged down with config details
+* Align naming and return types for affected code
+* Consolidate `SetterConfig` access/logic to `ackgenconfig` and/or `ackmodel`
+
+
+#### Out of Scope
+* Changing logic/algorithms
+  * the goal is minimal impact to functionality; once everything is in its proper place we can address inefficiencies separately
+* Consistent styling/naming throughout codebase
+
+### Test plan
+* add "sufficient" testing beforehand for quick feedback
+* update existing unit tests to reflect any changed logic due to migration
+* run a "healthy" subset of service controller e2e tests for validation
+
+## Appendix
+
+### Long paramaters
+The majority of functions in `code` package take both `ackgenconfig` and `ackresource` as parameters, but this is unnecessary because [`ackresource` has `ackgenconfig` as an attribute](https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/model/crd.go#L65):
```

Comment:
Can you change these links to perma links so that they are relevant after any changes to the code in those files.